src/Npgsql/ClusterAwareDataSource.cs

```
@@ -0,0 +1,489 @@
+    /// <summary>
+    /// Connection settings
+    /// </summary>
+    protected NpgsqlConnectionStringBuilder settings;
+    internal NpgsqlDataSourceConfiguration dataSourceConfig;
+
+    /// <summary>
+    /// List of initial hosts for control connections
+    /// </summary>
+    protected List<string>? initialHosts;
+
+    /// <summary>
+    /// Lock object
+    /// </summary>
+    protected static readonly object lockObject = new object();
+
+    internal ClusterAwareDataSource(NpgsqlConnectionStringBuilder settings, NpgsqlDataSourceConfiguration dataSourceConfig, bool useClusterAwareDataSource)
+        : base(settings, dataSourceConfig)
+    {
+        this.settings = settings;
+        this.dataSourceConfig = dataSourceConfig;
+        REFRESH_LIST_SECONDS = settings.YBServersRefreshInterval > 0 && settings.YBServersRefreshInterval < MAX_REFRESH_INTERVAL
+            ? this.settings.YBServersRefreshInterval
+            : 300;
+        initialHosts = Settings.Host?.Split(',').ToList();
+        Debug.Assert(initialHosts != null, nameof(initialHosts) + " != null");
+        if(useClusterAwareDataSource)
+        {
+            instance = this;
+            foreach (var host in initialHosts.ToList())
+            {
+                try
+                {
+                    var controlSettings = settings;
+                    controlSettings.Host = host.ToString();
+                    NpgsqlDataSource control = new UnpooledDataSource(controlSettings, dataSourceConfig);
+                    NpgsqlConnection controlConnection = NpgsqlConnection.FromDataSource(control);
+                    controlConnection.Open();
+                    CreatePool(controlConnection);
+                    controlConnection.Close();
+                    break;
+                }
+                catch (Exception)
+                {
+                    initialHosts.Remove(host);
+                    if (initialHosts.Count == 0)
+                        throw;
```

Should throw an exception with proper message.

---

This rethrows the caught exception